Support new intermine webservices

heralden opened the pull request:
This PR introduces many new functions for the new intermine webservices.

I'll be linking bluegenes PRs that rely on these new services, so the features can be easily tested.

heralden committed:
Implement save/im-list-remove-tag

heralden committed:
Default to identity xform in clj (to match cljs behavior)

heralden committed:
Implement auth/login service for logging in and merging lists

heralden committed:
Implement auth/logout service for invalidating token

heralden committed:
Implement save/im-list-update for updating list descriptions

I also had to add restful support for :put requests.

heralden committed:
Implement auth/change-password for setting new password

heralden committed:
Autoformat code

heralden committed:
Implement auth/register for registering a new user

yochannah commented on src/cljs/imcljs/internal/io.cljs:
```diff
@@ -40,6 +41,16 @@
             (wrap-auth token))))
 
+(defn put-wrapper-
+  "Returns the results of queries as table rows."
```
Is this true, or possibly copypasta? 😆

heralden replied:
The same docstring is present on all the wrappers. I don't know what's real anymore!

yochannah commented on src/clj/imcljs/internal/io.clj:
```diff
@@ -69,17 +70,31 @@
+  (parse-response (or xform identity)
+                  (client/post (url root path)
+                               (-> options ; Blank request map
+                                  ;(wrap-accept)
```
Am I right in thinking these commented lines are legacy stuff that should probably be deleted? If so, let's not copy it over to new methods.

yochannah commented on src/cljc/imcljs/internal/defaults.cljc:
```diff
@@ -40,6 +40,12 @@
                                   (missing? options :format) (assoc :format "json")))
     request))
 
+(defn wrap-put-defaults [request options]
```
it would be nice to add a docstring here, even if it's terse - I can see this is a wrapper function but we could describe what it wraps or what it is used for/by :)

yochannah reviewed (commented):
I'll test the functionality imminently - in the meantime a few small comments: 

1. I think there are old legacy commented out code lines. We don't have to go through and clean them all up right now, but at a minimum we shouldn't port them to new methods like the PUT function.  
2. A few functions are missing docstrings. They don't have to be long - \"(defn register
  "Register a new user account." ...` is great.
3. imcljs has some tests built in - let's add tests for any new fucntionality where practical.

heralden commented:
Tests are a bit tricky since Flymine (which is being tested against) doesn't have the new webservices yet.

I want to modify the imcljs tests to use a local biotestmine built with https://github.com/intermine/ci-scripts instead, although that will require changes in all the test files (and some more work if any of the tests relies on specific data being present on flymine).

I wonder if it would be better to have this as a separate PR?

yochannah commented:
Yep, that sounds fair!

heralden commented:
Regarding points 1 and 2, I can't bring myself to fix only the function I added and have it stick out like a sore thumb in the source file (as the rest of the file has the same problem). ]=

I'd rather fix the whole source file at a time, but I feel this should be its own PR.

Same with the legacy code, although that file (the CLJ end of `internal.io`) really needs some TLC, as it's missing some HTTP methods and has repeating logic (the requests it generate work but have data duplicated across the map).
Actually, I'd prefer to scrap that file and merge it with `io.cljs` into `io.cljc`, as I think this would be cleaner and fix the problem of the CLJ code not being updated due to not being used/tested.

So how about creating separate issues for these code smell problems?

heralden committed:
Fix put-body-wrapper- calling post instead of put

yochannah commented:
thumbs up from here!

heralden commented:
Coolio. I created the issue https://github.com/intermine/imcljs/issues/44 for this.

I believe all that's left for this PR is to add tests for the new webservices once https://github.com/intermine/imcljs/pull/43 gets merged.

heralden committed:
Merge remote-tracking branch 'upstream/dev' into feature/new-services

heralden committed:
Tests for imcljs.auth and new list webservices

heralden commented:
The tests are failing on Travis because it doesn't use the development build of Intermine, which has the new webservices. Currently working on a practical way to run the development build in CI without having to build it.

heralden committed:
fetch/preferences and save/preferences with tests

heralden committed:
auth/deregistration and auth/delete-account with tests

heralden committed:
save/delete-preference with tests added to runner

heralden committed:
fetch/saved-queries and save/query,delete-query with tests

heralden committed:
Sterilize with enforce-sort-order and vectorize enforce-sorting

We should use a vector instead of a list to match how we want sort
priority in BlueGenes to work.

heralden committed:
Update auth/change-password to require oldPassword argument

heralden committed:
Make path/walk handle path vector with string values

yochannah commented on src/cljc/imcljs/query.cljc:
```diff
@@ -97,10 +97,19 @@
                             (conj total (assoc constraint :code next-available-code))))) [] constraints)))
     query))
 
+(defn enforce-sort-order
```
nice one here! 👍

yochannah reviewed (commented):
The code looks great! needs a little more in the way of docs before I can merge, I think - or possibly I'm set up wrong in some way. I'd have added this myself but I wasn't sure what I needed to do to change things and make the tests run :) 

~I ran this against a biotestmine and against alpha.flymine, but both times I get auth-related failures. This is the output - any suggestions? Once I have enough info to get the tests passing I am happy to merge.~ 

Edit - I was indeed doing something wrong. Hold on for now.
